---
output: html_document
editor_options: 
  chunk_output_type: console
---

# Supplement: Autocorrelation among Markers
In this analysis, we will look at the autocorrelation (as a proxy of linkage disequilibrium) among our SNPs. 

```{r, include=FALSE}
knitr::opts_chunk$set(echo = T, warning = F, message = F, eval = T, results = 'hide',
                      fig.align = 'center', fig.width = 8, fig.height = 8)
knitr::opts_knit$set(root.dir = here::here())
```
```{r}
library(tidyverse)
library(vcfR)
library(vcfRmanip)
```

```{r}
vcf.DRC <- vcfR::read.vcfR("data/derived_data/bigbarcode_genetic_data/mipbivcfR.DRC.vcf")
LDcalc_vcf_DRC <- vcfRmanip::genautocorr(vcfR = vcf.DRC, biallelicsnps = T)
```

```{r}
# extract these and then use ggcor mat going to have facet pretty long since we have 14
LDcorrmat <- purrr::map(LDcalc_vcf_DRC, "corMat")
# lift over
LDcorrmatdf <- lapply(LDcorrmat, function(x){
  ret <- as.dist(x)
  ret <- broom::tidy(ret)
  return(ret)
})
names(LDcorrmatdf) <- names(LDcalc_vcf_DRC)
# bind cor mat 
LDcorrmatdf <- LDcorrmatdf %>% 
  dplyr::bind_rows(., .id = "CHROM")

# bring in position information
CHROMPOS <- cbind.data.frame(CHROM = vcfR::getCHROM(vcf.DRC),
                             POS = vcfR::getPOS(vcf.DRC))
CHROMPOS <- CHROMPOS %>% 
  dplyr::group_by(CHROM) %>% 
  dplyr::mutate(index = seq(1:length(CHROM)))

# now lift that over
colnames(CHROMPOS)[3] <- "item1" 
LDcorrmatdf <- dplyr::left_join(LDcorrmatdf, CHROMPOS, by = c("CHROM", "item1"))
colnames(CHROMPOS)[3] <- "item2" 
LDcorrmatdf <- dplyr::left_join(LDcorrmatdf, CHROMPOS, by = c("CHROM", "item2"))

```


### Overview of Genetic Information

#### Number of sites 

```{r, results='asis'}
CHROMPOS.num <- CHROMPOS %>% 
  dplyr::group_by(CHROM) %>% 
  dplyr::summarise(n = n())

DT::datatable(CHROMPOS.num, extensions='Buttons',
               options = list(
              searching = T,
              pageLength = 14,
              dom = 'Bfrtip', 
              buttons = c('csv')))

```

#### Autocorrelation
```{r}

chromposlvls <- paste0(CHROMPOS$CHROM, "_", CHROMPOS$POS)

LDcorrmatdf <- LDcorrmatdf %>% 
  dplyr::mutate(CHROMPOS.x = paste0(CHROM, "_", POS.x),
                CHROMPOS.y = paste0(CHROM, "_", POS.y),
                CHROMPOS.x = factor(CHROMPOS.x, levels = chromposlvls),
                CHROMPOS.y = factor(CHROMPOS.y, levels = chromposlvls))

LDcorrmatdf.plotObj <- LDcorrmatdf %>% 
    dplyr::mutate(CHROMPOS.x = forcats::fct_rev(forcats::fct_reorder(.f = CHROMPOS.x, .x = CHROMPOS.x, .fun = length)),
                CHROMPOS.y = forcats::fct_rev(forcats::fct_reorder(.f = CHROMPOS.y, .x = CHROMPOS.y, .fun = length))) %>%
  ggplot() + 
  geom_tile(aes(x = CHROMPOS.x, y = CHROMPOS.y, fill = distance)) +
  scale_fill_viridis_c("Pearson Correlation") + 
  xlab("Genomic Position") + ylab("Genomic Position") +
  facet_wrap(. ~ CHROM, shrink=F, scales = "free") +
  theme_minimal() + 
  theme(
    axis.title.y = element_text(vjust = 0.5, hjust = 0.5, size = 11, face = "bold"), 
    axis.title.x = element_text(vjust = 0.5, hjust = 0.5, size = 11, face = "bold"),
    axis.text.y = element_blank(),
    axis.text.x = element_blank(),
    legend.position = "right",
    legend.text = element_text(hjust = 0.5, vjust = 0.5, size = 8),
    plot.background = element_blank(),
    panel.grid = element_blank()
    )

jpgsnapshot(outpath = "~/Documents/GitHub/Space_the_Final_Miptier/results/figures/LD_snp_correlationplot.jpg",
            plot = LDcorrmatdf.plotObj)

```
```{r, results='asis'}
plot(LDcorrmatdf.plotObj)

```

```{r, results='asis'}

LDcorrmatdf %>% 
  dplyr::group_by(CHROM) %>% 
  dplyr::summarise(
    ncompar = n(),
    mean_corr = mean(distance)
  ) %>% 
  dplyr::mutate(mean_corr_round = round(mean_corr, 2)) %>% 
  DT::datatable(., extensions='Buttons',
               options = list(
              searching = T,
              pageLength = 14,
              dom = 'Bfrtip', 
              buttons = c('csv')))

```



